snitun/multiplexer/transport.py

```diff
@@ -0,0 +1,189 @@
+"""An asyncio.Transport implementation for multiplexer channel."""
+
+from __future__ import annotations
+
+import asyncio
+from asyncio import Transport
+import asyncio.sslproto
+import logging
+import sys
+
+from ..exceptions import MultiplexerTransportClose
+from ..multiplexer.channel import MultiplexerChannel
+from ..utils.asyncio import create_eager_task
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _feed_data_to_buffered_proto(proto: asyncio.BufferedProtocol, data: bytes) -> None:
+    """Feed data to a buffered protocol.
+
+    Adapted from asyncio.protocols
+    https://github.com/python/cpython/blob/c1f352bf0813803bb795b796c16040a5cd4115f2/Lib/asyncio/protocols.py#L200-L226
+
+    This function is used to feed data to a buffered protocol. It is
+    used to handle the case where the protocol's buffer is smaller than
+    the data to be fed to it.
+    """
+    data_len = len(data)
+    while data_len:  # pragma: no branch
+        buf = proto.get_buffer(data_len)
+        buf_len = len(buf)
+        if not buf_len:
+            raise RuntimeError("get_buffer() returned an empty buffer")
+
+        if buf_len >= data_len:
+            buf[:data_len] = data
+            proto.buffer_updated(data_len)
+            return
+
+        buf[:buf_len] = data[:buf_len]
+        proto.buffer_updated(buf_len)
+        data = data[buf_len:]
+        data_len = len(data)
+
+
+class ChannelTransport(Transport):
+    """An asyncio.Transport implementation for multiplexer channel."""
+
+    _start_tls_compatible = True
+
+    def __init__(self, channel: MultiplexerChannel) -> None:
+        """Initialize ChannelTransport."""
+        self._channel = channel
+        self._loop = asyncio.get_running_loop()
+        self._protocol: asyncio.BufferedProtocol | None = None
+        self._pause_future: asyncio.Future[None] | None = None
+        self._reader_task: asyncio.Task[None] | None = None
+        super().__init__(extra={"peername": (str(channel.ip_address), 0)})
```

Comment:
Until the production network is changed to pass the `ip_address` of the client and not the multiplexer, changing this to `127.0.0.1` will preserve the existing behavior

Reply:
Need to chat with @mdegat01 on this topic before this can be merged

Reply:
Adjusted to be backwards compat

Added comment in `Deployment considerations` above on how to make it pass the IP